EngineLayer/GlycoSearch/GlycoSearchEngine.cs

```diff
@@ -0,0 +1,516 @@
+﻿using EngineLayer.ModernSearch;
+using MzLibUtil;
+using Proteomics;
+using Proteomics.Fragmentation;
+using Proteomics.ProteolyticDigestion;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using EngineLayer;
+using MassSpectrometry;
+
+namespace EngineLayer.GlycoSearch
+{
+    public class GlycoSearchEngine : ModernSearchEngine
```

Comment:
i ran this task on your test data and it worked and i got glycopeptide results out. i was able to put the psm file into metadraw and view results. so that is good news.

the main issue that i have is this is 5600 lines of code and i really don't have a reasonable understanding how any of it works. i think this PR needs some high-level comments at the top of each new class (and ideally at the top of each new method) to explain what is going on here, since i think you're graduating soon and it will be up to us to maintain this code. there are also many variables and numbers that i don't understand the meaning of (i've commented a few as examples). can you work with @zrolfs and @trishorts to comment these files to the point where they can understand and explain what this code does?


some other things i noticed when i ran the glyco search:
- results.txt/allResults.txt is empty
- PEP/PEP-qvalue are all 0 for single_psm.tsv
- seen_glyco_localization.tsv lists several -1 values under "Max Site Specific Probability"
- seen_glyco_localization.tsv - accession (first column) is spelled incorrectly (missing the n in accession)
- selecting glycans to search for - no checkboxes to select glycans? need to look in/edit text file to see what is in DB - this is not very user-friendly and doesn't really behave like the rest of metamorpheus

Reply:
Thank you! I will try to add more comments. I think Zach and Shortreed get most of the parts.

Previously I wrote some Brief overview below (maybe I should put it somewhere in MM. )

Basicly, I create a new Glycan-Search task, as the Crosslink-Search task I did before. The reason why I open a new task is mainly to handle the complexity of glycoproteomics. Also, the code in this pull request is separated, so it won't hurt any other functionalities in MM.

There are four parts of the pull request: GlycoSearch GUI, GlycoSearch Task, GlycoSearch Engine, GlycoSearch Unit test.
GlycoSearch GUI and GlycoSearch Task are highly similar to Crosslink ones. GlycoSearch Task has a few functions specified. For GUI part, you don't want to spend too much time on it.

GlycoSearch Engine is the core. in EngineLayer -> GLycoSearch.
There are 5 classes handle glycan database: Node, Glycan, ModBox, GlycanBox, GlycanDatabase. GlycanDatabase read the glycans and ModBox GlycanBox will combine O-glycans into groups. Glycan and Node are basic classes for handling glycan masses, compostions and structures.
LocalizaitonGraph is used for graph localization. Glycans grouped by ModBox and GlycanBox will be used for localization. The main logic is in function ''LocalizeOGlycan".
GlycoSearchEngine is the core to work with the searching. The function "FindOGlycopeptideHashLocal" contains the logic.
GlycoSpectralMatch is glyco-psms. GlycoPeptides contains some functions for glycopeptide fragmentations, oxonium ion detection etc.
GlycoSearch Test could be used to help running through the code.

Basicly, I think it will be easy to start from GlycoSearchEngine.cs. There, it will load the glycan database, run ion-index open search, oxonium ion filter, graph localization, and create gsms (glyco-psms). Then it is easy to track other functions.
Thanks again. This pull request will be further improved by your effect.

Reply:
- results.txt/allResults.txt is empty. I will see what should be added there. Thanks.
- PEP is not calucated, but MetaDraw required PEP.
- seen_glyco_localization.tsv lists several -1 values, which is expected. I will explain this in WiKi. Thanks. 
- Thanks for pointing out the typo.
- I assume you mean to treat glycans as modifications in the MM. But the whole glycan world use a totally different format which is more flexible. I guess I will keep it like this, which is what the other software programs did.